File: raspberry-pi-ios-client/RaspberryPiClient/Data/Infrastructure/Foxglove/Common/FoxgloveMessageOperation.swift
//
//  FoxgloveMessageOperation.swift
//  RaspberryPiClient
//
//  Created by Shusuke Ota on 2025/03/20.
//

import Foundation

/// Foxglove WebSocket プロトコルの JSON メッセージ種別（JSONの"op"フィールド）
/// Call Service / Topic Publish / Subscribe に必要な op のみサポート。それ以外は .other として受信
/// Ref: https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md
enum FoxgloveMessageOperation: String, Codable {
    // Server → Client
    case serverInfo = "serverInfo"
    case advertise
    case unadvertise
    case advertiseServices = "advertiseServices"
    case unadvertiseServices = "unadvertiseServices"
    case serviceCallFailure = "serviceCallFailure"

    // Client → Server
    case subscribe
    case unsubscribe

    /// 未サポートの op（parameterValues, fetchAsset 等）
    case other = "__other__"

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        let raw = try container.decode(String.self)
        if let known = FoxgloveMessageOperation(rawValue: raw) {
            self = known
        } else {
            self = .other
        }
    }
}

File: raspberry-pi-android-client/app/src/main/java/com/shusuke/raspberry_pi_android_client/data/infrastructure/foxglove/common/FoxgloveMessageOperation.kt
package com.shusuke.raspberry_pi_android_client.data.infrastructure.foxglove.common

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Foxglove WebSocket プロトコルの JSON メッセージ種別（op フィールド）
 *
 * [wireValue] は JSON の "op" 文字列と一致する（[fromWireString] で逆引きする）。
 *
 * Ref: https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md
 */
@Serializable
enum class FoxgloveMessageOperation(val wireValue: String) {
    @SerialName("serverInfo")
    ServerInfo("serverInfo"),

    @SerialName("advertise")
    Advertise("advertise"),

    @SerialName("unadvertise")
    Unadvertise("unadvertise"),

    @SerialName("advertiseServices")
    AdvertiseServices("advertiseServices"),

    @SerialName("unadvertiseServices")
    UnadvertiseServices("unadvertiseServices"),

    @SerialName("serviceCallFailure")
    ServiceCallFailure("serviceCallFailure"),

    @SerialName("subscribe")
    Subscribe("subscribe"),

    @SerialName("unsubscribe")
    Unsubscribe("unsubscribe");

    companion object {
        private val wireValueToOperation: Map<String, FoxgloveMessageOperation> =
            entries.associateBy { it.wireValue }

        /**
         * JSON の "op" 文字列から [FoxgloveMessageOperation] を解決する。
         * 未知の値は null。
         */
        fun fromWireString(value: String): FoxgloveMessageOperation? =
            wireValueToOperation[value]
    }
}
